signature/cose/envelope.go

```diff
@@ -0,0 +1,576 @@
+	if err := generateProtectedHeaders(req, msg.Headers.Protected); err != nil {
+		return nil, &signature.InvalidSignRequestError{Msg: err.Error()}
+	}
+
+	// generate payload of COSE envelope
+	msg.Headers.Protected[cose.HeaderLabelContentType] = req.Payload.ContentType
+	msg.Payload = req.Payload.Content
+
+	// core sign process, generate signature of COSE envelope
+	if err := msg.Sign(rand.Reader, nil, signer); err != nil {
+		return nil, &signature.InvalidSignRequestError{Msg: err.Error()}
+	}
+
+	// generate unprotected headers of COSE envelope
+	generateUnprotectedHeaders(req, signer, msg.Headers.Unprotected)
+
+	// TODO: needs to add headerKeyTimeStampSignature.
+
+	// encode Sign1Message into COSE_Sign1_Tagged object
+	encoded, err := msg.MarshalCBOR()
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	e.base = msg
+	return encoded, nil
+}
+
+// Verify implements signature.Envelope interface.
+// Note: Verfiy only verifies integrity of the given COSE envelope
+func (e *envelope) Verify() (*signature.EnvelopeContent, error) {
+	// sanity check
+	if e.base == nil {
+		return nil, &signature.SignatureEnvelopeNotFoundError{}
+	}
+
+	certs, ok := e.base.Headers.Unprotected[cose.HeaderLabelX5Chain].([]interface{})
+	if !ok || len(certs) == 0 {
+		return nil, &signature.InvalidSignatureError{Msg: "certificate chain is not present"}
+	}
+	certRaw, ok := certs[0].([]byte)
+	if !ok {
+		return nil, &signature.InvalidSignatureError{Msg: "COSE envelope malformed leaf certificate"}
+	}
+	cert, err := x509.ParseCertificate(certRaw)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: "malformed leaf certificate"}
+	}
+
+	// core verify process, verify integrity of COSE envelope
+	publicKeyAlg, err := getSignatureAlgorithm(cert)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	verifier, err := cose.NewVerifier(publicKeyAlg, cert.PublicKey)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	err = e.base.Verify(nil, verifier)
+	if err != nil {
+		return nil, &signature.SignatureIntegrityError{Err: err}
+	}
+
+	// extract content
+	return e.Content()
+}
+
+// Content implements signature.Envelope interface.
+func (e *envelope) Content() (*signature.EnvelopeContent, error) {
+	// sanity check
+	if e.base == nil {
+		return nil, &signature.SignatureEnvelopeNotFoundError{}
+	}
+
+	payload, err := e.payload()
+	if err != nil {
+		return nil, err
+	}
+	signerInfo, err := e.signerInfo()
+	if err != nil {
+		return nil, err
+	}
+	return &signature.EnvelopeContent{
+		SignerInfo: *signerInfo,
+		Payload:    *payload,
+	}, nil
+}
+
+// Given a COSE envelope, extracts its signature.Payload.
+func (e *envelope) payload() (*signature.Payload, error) {
+	cty, ok := e.base.Headers.Protected[cose.HeaderLabelContentType]
+	if !ok {
+		return nil, &signature.InvalidSignatureError{Msg: "missing content type"}
```

Comment:
I feel `invalid content type ` or `unsupported content type` might make more sense.

Reply:
Line 272 can only check if cose.HeaderLabelContentType exist, it does not validate it.